From 8a94a8cb25087bc2287f2fdb71dd6e0fb22f53a9 Mon Sep 17 00:00:00 2001
From: Weijie Gao <weijie.gao@mediatek.com>
Date: Wed, 19 Nov 2025 08:56:26 +0800
Subject: [PATCH 13/23] board: mediatek: fix incorrect ubi partition recovery

Before erasing the ubi partition, the whole ubi must be unloaded to remove
any previous error states.

Change-Id: I380bcab9f0739014d1d9975a1f71009e16dbe46c
Signed-off-by: Weijie Gao <weijie.gao@mediatek.com>
CR-Id: WCNCR00180092
---
 board/mediatek/common/mtd_helper.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/board/mediatek/common/mtd_helper.c b/board/mediatek/common/mtd_helper.c
index 949ce0a3aa4..adf28337a9a 100644
--- a/board/mediatek/common/mtd_helper.c
+++ b/board/mediatek/common/mtd_helper.c
@@ -677,12 +677,14 @@ static int mount_ubi(struct mtd_info *mtd, bool create)
 			cprintln(CAUTION, "*** Failed to attach UBI ***");
 			cprintln(NORMAL, "*** Rebuilding UBI ***");
 
+			ubi_detach();
+
 			ret = mtd_erase_skip_bad(mtd, 0, mtd->size, mtd->size,
 						 NULL, NULL, NULL, false);
 			if (ret)
 				return ret;
 
-			ret = ubi_init();
+			ret = ubi_part(mtd->name, NULL);
 		}
 
 		if (ret) {
-- 
2.45.2

